src/Service/FileWatcher/Service_FileWatcher.re

```diff
@@ -0,0 +1,76 @@
+open Oni_Core;
+
+module Log = (val Log.withNamespace("Oni2.Service.FileWatcher"));
+
+[@deriving show({with_path: false})]
+type event = {
+  path: string,
+  hasRenamed: bool,
+  hasChanged: bool,
+};
+
+module WatchSubscription =
+  Isolinear.Sub.Make({
+    type state = {
+      path: string,
+      watcher: option(Luv.FS_event.t),
+    };
+
+    type nonrec msg = event;
+    type nonrec params = string;
+
+    let name = "FileWatcher";
+    let id = Fun.id;
+
+    let init = (~params as path, ~dispatch) => {
+      Log.tracef(m => m("Starting file watcher for %s", path));
+
+      switch (Luv.FS_event.init()) {
+      | Ok(watcher) =>
+        let onEvent = (
+          fun
+          | Ok((_file, events)) =>
+            dispatch({
+              path,
+              hasRenamed: List.mem(`RENAME, events),
+              hasChanged: List.mem(`CHANGE, events),
+            })
+          | Error(error) =>
+            Log.errorf(m => m("'%s': %s", path, Luv.Error.strerror(error)))
+        );
+
+        Luv.FS_event.start(watcher, path, onEvent);
+
+        {path, watcher: Some(watcher)};
+
+      | Error(error) =>
+        let message = Luv.Error.strerror(error);
+        Log.errorf(m => m("init failed for '%s': %s", path, message));
+
+        {path, watcher: None};
+      };
+    };
+
+    let update = (~params as path, ~state, ~dispatch as _) =>
+      if (path != state.path) {
+        // Should be unreachable because the path is the id
+        failwith(
+          "unreachable",
+        );
+      } else {
+        state;
+      };
```

Comment:
Why is the conditional necessary here? Would it make sense to just return `state`?

Reply:
Yeah, definitely. I started with just the `failtwith`, because I couldn't, and still can't, see any hood reason for update to get called if `params` hasn't changed. Then of course that crashed, so it morphed into this just to assure myself that it didn't do other weird things. But I'm pretty assured now, so no reason to keep it.